README.md
# Lelylan API Proxy

Lelylan API proxy.


## Requirements

API Proxy is tested against Node v0.10.36


## Installation

    $ git clone git@github.com:lelylan/api-proxy.git && cd api-proxy
    $ npm install && npm install -g foreman
    $ nf start


## Resources

* [Lelylan API](http://dev.lelylan.com/api)


## Contributing

Fork the repo on github and send a pull requests with topic branches.
This repo does not contain any test. If you wish to provide specs to your contribution let us know to say thank you.


## Coding guidelines

Follow [Felix](http://nodeguide.com/style.html) guidelines.


## Feedback

Use the [issue tracker](http://github.com/lelylan/api-proxy/issues) for bugs or [stack overflow](http://stackoverflow.com/questions/tagged/lelylan) for questions.
[Mail](mailto:dev@lelylan.com) or [Tweet](http://twitter.com/lelylan) us for any idea that can improve the project.


## Links

* [GIT Repository](http://github.com/lelylan/api-proxy)
* [Lelylan Dev Center](http://dev.lelylan.com)
* [Lelylan Site](http://lelylan.com)


## Authors

[Andrea Reginato](https://www.linkedin.com/in/andreareginato)


## Contributors

Special thanks to all [contributors](https://github.com/lelylan/api-proxy/contributors)
for submitting patches.


## Changelog

See [CHANGELOG](https://github.com/lelylan/api-proxy/blob/master/CHANGELOG.md)


## License

Lelylan is licensed under the [Apache License, Version 2.0](http://www.apache.org/licenses/LICENSE-2.0).


CHANGELOG.md
# Changelog

## v0.1.0 (2016.01.01)

First version Lelylan API Proxy


spec/server.spec.js
var nock    = require('nock');
var http    = require('http');
var request = require('request');

describe('when requires a resource', function() {

  var fixture = __dirname + '/fixtures/devices.json';
  var scope = nock('http://localhost:8000').get('/devices').replyWithFile('201', fixture);

  it('keeps status code', function(done) { 
    request('http://localhost:8000/devices', function(error, res, body) {
      expect(res.statusCode).toEqual('201');
      done();
    });
  });
});


app.js
var httpProxy = require('http-proxy');

var port = process.env.PORT || 8200;

// all not found routes are forwarded to the devices service
var routing = {
  '/me':            { port: process.env.LELYLAN_PEOPLE_PORT || 80,        host: process.env.LELYLAN_PEOPLE_URL },
  '/devices':       { port: process.env.LELYLAN_DEVICES_PORT || 80,       host: process.env.LELYLAN_DEVICES_URL },
  '/activations':   { port: process.env.LELYLAN_DEVICES_PORT || 80,       host: process.env.LELYLAN_DEVICES_URL },
  '/types':         { port: process.env.LELYLAN_TYPES_PORT || 80,         host: process.env.LELYLAN_TYPES_URL },
  '/properties':    { port: process.env.LELYLAN_TYPES_PORT || 80,         host: process.env.LELYLAN_TYPES_URL },
  '/functions':     { port: process.env.LELYLAN_TYPES_PORT || 80,         host: process.env.LELYLAN_TYPES_URL },
  '/statuses':      { port: process.env.LELYLAN_TYPES_PORT || 80,         host: process.env.LELYLAN_TYPES_URL },
  '/categories':    { port: process.env.LELYLAN_TYPES_PORT || 80,         host: process.env.LELYLAN_TYPES_URL },
  '/subscriptions': { port: process.env.LELYLAN_SUBSCRIPTIONS_PORT || 80, host: process.env.LELYLAN_SUBSCRIPTIONS_URL },
}

var server = httpProxy.createServer(
  require('./lib/uri-middleware')(routing)
).listen(port);


lib/uri-middleware.js
function matcher (url, dest) {
  // First, turn the URL into a regex.
  // NOTE: Turning user input directly into a Regular Expression is NOT SAFE.
  var r = new RegExp(url.replace(/\//, '\\/'));
  // This next block of code may look a little confusing.
  // It returns a closure (anonymous function) for each URL to be matched,
  // storing them in an array - on each request, if the URL matches one that has
  // a function stored for it, the function will be called.
  return function (url) {
    var m = r.exec(url)
    if (!m) {
      return;
    }
    // keep the path as we want to change only the domain
    var path = url;
    //console.log('proxy:', url, '->', dest);
    return {url: path, dest: dest};
  }
}

module.exports = function (urls) {
  // This is the entry point for our middleware.
  // 'matchers' is the array of URL matchers, as mentioned above.
  var matchers = [];
  for (var url in urls) {
    // Call the 'matcher' function above, and store the resulting closure.
    matchers.push(matcher(url, urls[url]));
  }

  // This closure is returned as the request handler.
  return function (req, res, next) {
    //
    // in node-http-proxy middlewares, `proxy` is the prototype of `next`
    // (this means node-http-proxy middlewares support both the connect API (req, res, next)
    // and the node-http-proxy API (req, res, proxy)
    //
    var proxy = next;
    for (var k in matchers) {
      // for each URL matcher, try the request's URL.
      var m = matchers[k](req.url);
      // If it's a match:
      if (m) {
        // Replace the local URL with the destination URL.
        req.url = m.url;
        // If routing to a server on another domain, the hostname in the request must be changed.
        req.headers['x-host'] = process.env.LELYLAN_PROXY_URL;
        req.headers.host = m.dest.host;
        // Once any changes are taken care of, this line makes the magic happen.
        return proxy.proxyRequest(req, res, m.dest);
      }
    }

    // default redirect when no routes matches
    // TODO refactor this part of code as it is duplicated
    var m = { url: '/devices', dest: { port: process.env.LELYLAN_DEVICES_PORT || 80,  host: process.env.LELYLAN_DEVICES_URL } }
    req.url = m.url;
    req.headers['x-host'] = process.env.LELYLAN_PROXY_URL;
    req.headers.host = m.dest.host;
    return proxy.proxyRequest(req, res, m.dest);

    // TODO use it if you want to use other middlewares
    //next() //if there was no matching rule, fall back to next middleware.
  }
}


web.js
var httpProxy = require('http-proxy');

var port = process.env.PORT || 8000;

var routing = {
  '/devices':       { port: process.env.DEVICES_PORT || 80,      host: process.env.DEVICES_URI },
  '/consumptions':  { port: process.env.DEVICES_PORT || 80,      host: process.env.DEVICES_URI },
  '/histories':     { port: process.env.DEVICES_PORT || 80,      host: process.env.DEVICES_URI },
  '/types':         { port: process.env.TYPES_PORT || 80,        host: process.env.TYPES_URI },
  '/properties':    { port: process.env.TYPES_PORT || 80,        host: process.env.TYPES_URI },
  '/functions':     { port: process.env.TYPES_PORT || 80,        host: process.env.TYPES_URI },
  '/statuses':      { port: process.env.TYPES_PORT || 80,        host: process.env.TYPES_URI },
  '/categories':    { port: process.env.TYPES_PORT || 80,        host: process.env.TYPES_URI },
  '/locations':     { port: process.env.LOCATIONS_PORT || 80,    host: process.env.LOCATIONS_URI },
  '/subscriptions': { port: process.env.SUBSCRIPTIONS_PORT || 80, host: process.env.SUBSCRIPTIONS_URI }
}

var server = httpProxy.createServer(
  require('./lib/uri-middleware')(routing)
).listen(port);
